src/sql/parts/objectExplorer/common/objectExplorerViewTreeShim.ts

```diff
@@ -47,73 +45,94 @@ export class OEShimService implements IOEShimService {
+		// we assume only nodes with payloads can be connected
+		// check to make sure we have an existing connection
+		let key = generateSessionMapKey(viewId, node);
+		let session = this.sessionMap.get(key);
+		if (session) {
+			let closed = (await this.oe.closeSession(node.childProvider, this.oe.getSession(session))).success;
+			if (closed) {
+				this.sessionMap.delete(key);
 			}
-			let sessionId = this.sessionMap.get(identifier).get(hash(node.payload || node.childProvider));
-			let treeNode = new TreeNode(undefined, undefined, undefined, node.handle, undefined, undefined, undefined, undefined, undefined, undefined);
-			let profile: IConnectionProfile = node.payload || {
-				providerName: node.childProvider,
-				authenticationType: undefined,
-				azureTenantId: undefined,
-				connectionName: undefined,
-				databaseName: undefined,
-				groupFullName: undefined,
-				groupId: undefined,
-				id: undefined,
-				options: undefined,
-				password: undefined,
-				savePassword: undefined,
-				saveProfile: undefined,
-				serverName: undefined,
-				userName: undefined,
-			};
-			treeNode.connection = new ConnectionProfile(this.capabilities, profile);
-			return TPromise.wrap(this.oe.resolveTreeNodeChildren({
-				success: undefined,
-				sessionId,
-				rootNode: undefined,
-				errorMessage: undefined
-			}, treeNode).then(e => e.map(n => this.mapNodeToITreeItem(n, node))));
-		} catch (e) {
-			return TPromise.as([]);
+			return closed;
+		}
+		return Promise.resolve(false);
+	}
+
+	private async getOrCreateSession(viewId: string, node: ITreeItem): Promise<string> {
+		// verify the map is correct
+		let key = generateSessionMapKey(viewId, node);
+		if (!this.sessionMap.has(key)) {
+			this.sessionMap.set(key, await this.createSession(viewId, node.childProvider, node));
 		}
+		return this.sessionMap.get(key);
 	}
 
-	private mapNodeToITreeItem(node: TreeNode, parentNode: ITreeItem): ITreeItem {
+	public async getChildren(node: ITreeItem, viewId: string): Promise<ITreeItem[]> {
+		let sessionId = await this.getOrCreateSession(viewId, node);
+		let requestHandle = this.nodeHandleMap.get(generateNodeMapKey(viewId, node)) || node.handle;
+		let treeNode = new TreeNode(undefined, undefined, undefined, requestHandle, undefined, undefined, undefined, undefined, undefined, undefined);
+		let profile: IConnectionProfile = node.payload || {
+			providerName: node.childProvider,
+			authenticationType: undefined,
+			azureTenantId: undefined,
+			connectionName: undefined,
+			databaseName: undefined,
+			groupFullName: undefined,
+			groupId: undefined,
+			id: undefined,
+			options: undefined,
+			password: undefined,
+			savePassword: undefined,
+			saveProfile: undefined,
+			serverName: undefined,
+			userName: undefined,
+		};
+		treeNode.connection = new ConnectionProfile(this.capabilities, profile);
+		return TPromise.wrap(this.oe.resolveTreeNodeChildren({
```

Comment:
you've been switching `TPromise` to `Promise` elsewhere, do you want to do that here?

Reply:
Aw, yes I should